### PR TITLE
Improve german verb form creation

### PR DESCRIPTION
**Pull request by manuelaugustin**
## Summary

This PR can be summarized in the following changelog entry:

* Makes it possible to create stems from German participles, e.g. `gekauft` -> `kauf`. From these stems, the correct verb and adjectives forms are then created.
* Adds functionality to create participle forms with prefixes, e.g. `eingekauft`.

## Relevant technical choices:

* The check for whether a word is a participle takes place after checking exceptions (since these also include participle stems) but before creating regular forms.
* If a participle is detected, we only create verb forms (assuming that the participle was used verbally, e.g. "er hat sich die Haare gefärbt" - "he dyed his hair") and adjective forms (assuming that the participle was used adjectivally, e.g. "die Haare sind gefärbt" - "the hair is dyed"). The adjective forms are based on the stem that has only the suffixes removed, not the prefixes. This is because we want forms such as "die gefärbten Haare" and not (incorrectly) "*die färbten Haare".

## Test instructions

This PR can be tested by following these steps:

* To check whether participles get stemmed correctly:
  * In the example, make sure your locale is set to `de_DE` and you have enabled morphology.
  * Add a text with at least 150 words.
  * Set a keyphrase containing a participle, e.g. `gekauft`.
  * In the text, add different forms of the verb `kaufen`, e.g. `kaufe`, `kaufte` etc.
  * Make sure these forms get recognized as forms of the keyphrase.
* To check whether participle forms get created correctly:
 * In the example, make sure your locale is set to `de_DE` and you have enabled morphology.
  * Add a text with at least 150 words.
  * Set a keyphrase containing a verb, e.g. `kaufen`.
  * In the text, add the participle `gekauft`.
  * Make sure the participle forms gets recognized as a form of the keyphrase.
* For different verb/verb forms for testing, consult a dictionary, e.g. https://de.wiktionary.org/wiki/Wiktionary:Hauptseite.

Moved from https://github.com/Yoast/YoastSEO.js/pull/2191

Fixes https://github.com/Yoast/YoastSEO.js/issues/2190

### CODE REVIEW

**Comment by manuelaugustin** [comment]
Implemented feedback from CR, see PR on old repo: https://github.com/Yoast/YoastSEO.js/pull/2191

File: packages/yoastseo/src/morphology/german/detectAndStemRegularParticiple.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * Detects whether a word is a regular participle without a prefix and if so, returns the stem.
+ *
+ * @param {Object}  morphologyDataVerbs The German morphology data for verbs.
+ * @param {string}  word                The word (not stemmed) to check.
+ *
+ * @returns {string|null} The stem or null if no participle was matched.
+ */
+const detectAndStemParticiplesWithoutPrefixes = function( morphologyDataVerbs, word ) {
+	const participleRegex1 = new RegExp( "^" + morphologyDataVerbs.participleStemmingClasses[ 1 ].regex );
+	const participleRegex2 = new RegExp( "^" + morphologyDataVerbs.participleStemmingClasses[ 0 ].regex );
```
Comment (hansjovis) [inline packages/yoastseo/src/morphology/german/detectAndStemRegularParticiple.js]: I would give these variables more descriptive names.
`geStemEtParticipleRegex` and `geStemTParticipleRegex`?

File: packages/yoastseo/src/morphology/german/generateParticipleForm.js [hunk 2]
```diff
@@ -18,10 +19,72 @@ const addParticipleAffixes = function( stemmedWord, affixes ) {
...
+
+			if ( stemmedWord.endsWith( "d" ) || stemmedWord.endsWith( "t" ) ) {
+				return addParticipleAffixes( stemmedWordWithoutPrefix, morphologyDataVerbs.participleAffixes.stemEndsInDOrT, currentPrefix );
+			}
+
+			return addParticipleAffixes( stemmedWordWithoutPrefix, morphologyDataVerbs.participleAffixes.regular, currentPrefix );
+		}
+	}
+
+	return "";
```
Comment (hansjovis) [inline packages/yoastseo/src/morphology/german/generateParticipleForm.js]: I would return `null` here, 
like in `detectAndStemParticiplesWithPrefixes` in `detectAndStemRegularParticiple.js`.

Makes it more apparent that no stem has been found.

File: packages/yoastseo/src/morphology/german/generateParticipleForm.js [hunk 3]
```diff
@@ -18,10 +19,72 @@ const addParticipleAffixes = function( stemmedWord, affixes ) {
...
+ * @returns {string} The created participle form.
+ */
+export function generateParticipleForm( morphologyDataVerbs, stemmedWord ) {
+	let participleFormWithPrefix = generateParticipleFormWithSeparablePrefix(
+		morphologyDataVerbs,
+		stemmedWord,
+		morphologyDataVerbs.verbPrefixesSeparable
+	);
+
+	if ( participleFormWithPrefix.length > 0 ) {
```
Comment (hansjovis) [inline packages/yoastseo/src/morphology/german/generateParticipleForm.js]:
```suggestion
	if ( participleFormWithPrefix ) {
```

File: packages/yoastseo/src/morphology/german/generateParticipleForm.js [hunk 4]
```diff
@@ -18,10 +19,72 @@ const addParticipleAffixes = function( stemmedWord, affixes ) {
...
+	 * For its these prefixes used in the inseparable form, the resulting participle would be the same as
+	 * the 3rd person singular, so we don't need to create a separate form here (e.g., "überführen" - "überführt").
+	 */
+	participleFormWithPrefix = generateParticipleFormWithSeparablePrefix(
+		morphologyDataVerbs,
+		stemmedWord,
+		morphologyDataVerbs.verbPrefixesSeparableOrInseparable
+	);
+
+	if ( participleFormWithPrefix.length > 0 ) {
```
Comment (hansjovis) [inline packages/yoastseo/src/morphology/german/generateParticipleForm.js]:
```suggestion
	if ( participleFormWithPrefix ) {
```

File: packages/yoastseo/src/morphology/german/detectAndStemRegularParticiple.js (hunk 1 above)
Comment (manuelaugustin) [reply packages/yoastseo/src/morphology/german/detectAndStemRegularParticiple.js]: Good point, that makes the code quite a bit easier to read!

**Comment by manuelaugustin** [comment]
Implemented the changes as suggested!

**Comment by hansjovis** [comment]
Acceptance test: ✅

**Pull request merged**